internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	needRelease := checkIfFSNeedQuiesceRelease(vgs, req.GetSourceVolumeIds())
+	if needRelease {
+		return cs.releaseQuiesceAndGetVolumeGroupSnapshotResponse(ctx, req, vgs, fsMap, vg, cr)
+	}
+
+	// If the volume group snapshot does not exist, reserve the volume group
+	if vgs == nil {
+		vgs, err = store.ReserveVolumeGroup(ctx, vg, cr)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to reserve volume group: %v", err)
+
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	inProgress, err := cs.queisceSubvolumes(ctx, req, vgs, fsMap, cr)
+	if err != nil {
+		log.ErrorLog(ctx, "failed to quiesce subvolumes: %v", err)
+		if err != nil && !errors.Is(err, cerrors.ErrInProgress) {
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
+			if uErr != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", uErr)
+			}
+		}
+
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if inProgress {
+		return nil, status.Error(codes.Internal, "Quiesce operation is in progress")
+	}
+
+	_, err = cs.createSnapshotAddToVolumeGroupJournal(ctx, req, false, vg, vgs, cr, fsMap, req.GetSecrets())
+	if err != nil {
+		log.ErrorLog(ctx, "failed to create snapshot and add to volume group journal: %v", err)
+
+		if !errors.Is(err, cerrors.ErrInProgress) {
+			log.ErrorLog(ctx, "failed to create snapshot and add to volume group journal: %v", err)
+			// Handle Undo reservation and timeout as well
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
+			if uErr != nil {
+				log.ErrorLog(ctx, "failed to delete snapshot and undo reservation: %v", uErr)
+			}
+
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.CreateVolumeGroupSnapshotResponse{}, nil
+}
+
+// queisceSubvolumes quiesces the subvolumes and subvolume groups present in
+// the filesystems of the volumeID's present in the
+// CreateVolumeGroupSnapshotRequest.
+func (cs *ControllerServer) queisceSubvolumes(ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+	cr *util.Credentials,
+) (bool, error) {
+	var inProgress bool
+	for _, fm := range fsMap {
+		// Quiesce the fs, subvolumes and subvolume groups
+		data, err := fm.SubvolumeQuiesce(ctx, vgs.RequestName)
+		if err != nil && !errors.Is(err, rados.ErrPermissionDenied) {
+			// ceph returns permission denied if the quiesce is already in
+			// progress if we execute the quiesce again until the QUIESCED is
+			// reached or any other error state is reached
+			log.ErrorLog(ctx, "failed to quiesce subvolume: %v", err)
+			uErr := cs.deleteSnapshotsAndUndoReservation(ctx, vgs, cr, fsMap, req.GetSecrets())
```

Comment:
Undoing the reservation here does not look like the right place. Ideally it should be done in the same function where the reservation was created (maybe in a cleanup helper function).